File: src/viewforge/router.py
# src/viewforge/router.py

from typing import Callable, Dict

class Router:
    def __init__(self, mode: str = "hash"):
        self.mode = mode
        self.routes: Dict[str, Callable[[], None]] = {}
        self.current_path = "/"
        self.window = None

    def register(self, path: str, handler: Callable[[], None]):
        self.routes[path] = handler

    def navigate(self, path: str):
        if self.window:
            if self.mode == "hash":
                self.window.evaluate_js(f'window.location.hash = "#{path}";')
            elif self.mode == "history":
                self.window.evaluate_js(
                    f'''
                    window.history.pushState(null, "", "/{path}");
                    window.dispatchEvent(new Event("popstate"));
                    '''
                )

    def attach_webview(self, webview_window):
        self.window = webview_window

File: static/js/router.js
// static/js/router.js

export function startRouter({ mode = 'hash', routes = {} }) {
    const getPath = () => {
        if (mode === 'hash') return window.location.hash.slice(1) || '/';
        return window.location.pathname || '/';
    };

    const render = () => {
        const path = getPath();
        const handler = routes[path];
        if (handler) {
            handler();
        } else {
            document.body.innerHTML = `<h1>404 Not Found: ${path}</h1>`;
        }
    };

    if (mode === 'hash') {
        window.addEventListener('hashchange', render);
    } else {
        window.addEventListener('popstate', render);
    }

    window.addEventListener('load', render);
}
